### ISSUE TITLE
🐛 False positive for `lint/correctness/noUnusedPrivateClassMembers` when using `??=`

### ISSUE THREAD

**Issue by lgarron** [biomejs/biome#7192]
### Environment information

```block
CLI:
  Version:                      2.1.4
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v24.5.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         bun/1.2.19

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  false

Workspace:
  Open Documents:               0
```

### What happened?

Repro link: https://biomejs.dev/playground/?indentStyle=space&indentWidth=4&bracketSameLine=true&code=YwBsAGEAcwBzACAAQQBwAHAAIAB7AAoAIAAgACMAcABlAHIAcwBpAHMAdABlAG4AYwBlAFIAZQBxAHUAZQBzAHQAOgAgAFAAcgBvAG0AaQBzAGUAPABiAG8AbwBsAGUAYQBuAD4AIAB8ACAAdQBuAGQAZQBmAGkAbgBlAGQAOwAKACAAIABzAGEAdgBlAEQAYQB0AGEAKAApACAAewAKACAAIAAgACAAdABoAGkAcwAuACMAcABlAHIAcwBpAHMAdABlAG4AYwBlAFIAZQBxAHUAZQBzAHQAIAA%2FAD8APQAgAG4AYQB2AGkAZwBhAHQAbwByAC4AcwB0AG8AcgBhAGcAZQAuAHAAZQByAHMAaQBzAHQAKAApADsACgAgACAAIAAgAC8ALwAgACYgCgAgACAAfQAKAH0ACgAKAG4AZQB3ACAAQQBwAHAAKAApADsA

Code:

```ts
class App {
  #persistenceRequest: Promise<boolean> | undefined;
  saveData() {
    this.#persistenceRequest ??= navigator.storage.persist();
    // …
  }
}

new App();
```

This results in a `lint/correctness/noUnusedPrivateClassMembers` error claiming that the `.#persistenceRequest` is field is unused. However, it is accessed on the line with `??=` to gate a call to a function with side effects.

In fact, the linter will no longer error if you change it to essentially equivalent code:

```ts
this.#persistenceRequest = this.#persistenceRequest ?? navigator.storage.persist();
```

### Expected result

No error.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by vladimir-ivanov** [comment]
@lgarron this is a valid failure, consistent with eslint, please see:


https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLXByaXZhdGUtY2xhc3MtbWVtYmVyczogXCJlcnJvclwiKi9cblxuY2xhc3MgQXBwIHtcbiAgI3BlcnNpc3RlbmNlUmVxdWVzdDtcbiAgc2F2ZURhdGEoKSB7XG4gICAgdGhpcy4jcGVyc2lzdGVuY2VSZXF1ZXN0ID8/PSBuYXZpZ2F0b3Iuc3RvcmFnZS5wZXJzaXN0KCk7XG4gIH1cbn1cbiIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==

**Comment by lgarron** [comment]
> [@lgarron](https://github.com/lgarron) this is a valid failure, consistent with eslint, please see:
> 
> https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLXByaXZhdGUtY2xhc3MtbWVtYmVyczogXCJlcnJvclwiKi9cblxuY2xhc3MgQXBwIHtcbiAgI3BlcnNpc3RlbmNlUmVxdWVzdDtcbiAgc2F2ZURhdGEoKSB7XG4gICAgdGhpcy4jcGVyc2lzdGVuY2VSZXF1ZXN0ID8/PSBuYXZpZ2F0b3Iuc3RvcmFnZS5wZXJzaXN0KCk7XG4gIH1cbn1cbiIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==

I don't follow. Why would `eslint`'s behaviour make this a valid failure?

**Comment by vladimir-ivanov** [comment]
because the assignment is never used anywhere else. Only if a return statement is added it will think that the intention is to be used elsewhere. Hope this makes sense?

**Comment by lgarron** [comment]
> because the assignment is never used anywhere else. Only if a return statement is added it will think that the intention is to be used elsewhere. Hope this makes sense?

No, it doesn't make sense to me. 😕

As I wrote above:

> This results in a `lint/correctness/noUnusedPrivateClassMembers` error claiming that the `.#persistenceRequest` is field is unused. However, it is accessed on the line with `??=` to gate a call to a function with side effects.
> 
> In fact, the linter will no longer error if you change it to essentially equivalent code:
> 
> this.#persistenceRequest = this.#persistenceRequest ?? navigator.storage.persist();

**Comment by vladimir-ivanov** [comment]
ok, so the bug is essentially -> this.#persistenceRequest = this.#persistenceRequest ?? navigator.storage.persist();

**Comment by vladimir-ivanov** [comment]
https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLXByaXZhdGUtY2xhc3MtbWVtYmVyczogXCJlcnJvclwiKi9cblxuY2xhc3MgQXBwIHtcbiAgI3BlcnNpc3RlbmNlUmVxdWVzdDtcbiAgc2F2ZURhdGEoKSB7XG4gICAgdGhpcy4jcGVyc2lzdGVuY2VSZXF1ZXN0ID0gdGhpcy4jcGVyc2lzdGVuY2VSZXF1ZXN0ID8/IG5hdmlnYXRvci5zdG9yYWdlLnBlcnNpc3QoKTtcbiAgfVxufVxuIiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19 this also seems consistent with eslint - no error

**Comment by vladimir-ivanov** [comment]
Personally I tend to agree with you, but the rule suggest we do the same behaviour as eslint:        

sources: &[RuleSource::Eslint("no-unused-private-class-members").same()],

**Comment by vladimir-ivanov** [comment]
to me the second case should also be marked as invalid - possibly should be reported as bug on eslint :-)
@arendjr, @ematipico  your opinion on this please?

**Comment by lgarron** [comment]
I don't mean to be facetious, but I don't understand why I should be concerned about `eslint`'s behaviour. I use Biome specifically because I do not want to use `eslint`.

This is clearly an access of `this.#persistenceRequest`, as is allowed by Biome:

```js
this.#persistenceRequest = this.#persistenceRequest ?? navigator.storage.persist();
```

The access of `this.#persistenceRequest` determines whether to run code with side effect, which sounds very clearly like usage to me. I think it would only make sense to error if the right expression was proven to have no side effects, but I wouldn't expect Biome to implement that analysis.

And this is effectively equivalent[^1], which is why I am very surprised to see is marked with an error:

```js
this.#persistenceRequest ??= navigator.storage.persist();
```

I would also strongly prefer to use the latter code without having to ignore a warning, as it is more idiomatic, shorter, and avoids the chance for maintenance mistakes due to a duplicated expression.

[^1]: I say "effectively" because programmers will generally be able to use them with the same semantics in most programs and `??=` exists in large part to enable the former expression to be replaced with the latter. Technically, it's possible to implement arbitrary behaviour if the property access is defined with a getter and setter. But I think in that case I would definitely expect it to be beyond Biome's power to correctly declared the field "unused".

**Comment by lgarron** [comment]
Going even further, if these are considered errors I'm not sure where the line would even make sense to draw.

This code could further be expanded into:

```js
this.#persistenceRequest = this.#persistenceRequest ? this.#persistenceRequest : navigator.storage.persist();
```

I don't really see how it could be argued that `this.#persistenceRequest` is not "used" here.

**Comment by ematipico** [comment]
> but the rule suggest we do the same behaviour as eslint:

That doesn't mean that eslint doesn't have bugs or doesn't cover specific cases ;) and if we want to diverge from the rule, we can do it whenever we want at will, as long as we document the cases we diverge/cover. Those aren't hard requirements, they are more soft guidelines.

Also, you've answered your question. If you think it's a bug, then let's improve the rule. And if we diverge in the process, so be it 🫡

**Comment by vladimir-ivanov** [comment]
so we have two options here -> 
BOTH are flagged as errors.
BOTH are flagged as valid.
My suggestion would be to mark both as invalid/ errors. that is so we are consistent with useReadonlyClassProperties:
e.g. from in https://github.com/biomejs/biome/blob/main/crates/biome_js_analyze/tests/specs/style/useReadonlyClassProperties/invalid.ts 

```ts
export class Test {
	private field: number;

	constructor() {
		this.field ??= 1;
	}
} 
```
@lgarron  please confirm if it makes sense to you? or if it does not :-) Maybe we should reevaluate useReadonlyClassProperties too (please go trough tests and make a proposal if so maybe)?

**Comment by vladimir-ivanov** [comment]
fun part is the useReadonlyClassProperties is fully consistent (all tests came from eslint themselves), yet they contradict on some areas with noUnusedPrivateClassMembers. Hm, been silly they are two separate offerings - one is eslint, the other is ts-eslint. Hence the inconsistencies.

**Comment by vladimir-ivanov** [comment]
class UsedMember {
	#usedMember = 42;
	foo() {
		this.#usedMember = this.#usedMember;
	}
} looking at current tests for no_unused_private_class_members this is marked as valid, so this.#persistenceRequest ??= navigator.storage.persist(); should ALSO be marked as valid, as @lgarron you suggested for consistency (with itself).

**Comment by vladimir-ivanov** [comment]
cool, sounds like there is no appetite for introducing consistency with useReadonlyClassProperties at the moment.
Then the scope of this task would be to make all of these:

```ts
export class Test {
	private field: number;

	constructor() {
		this.field ??= 1;
	}
} 
```

```ts
export class Test {
	private field: number;

	constructor() {
		this.field += 1;
	}
} 
```

```ts
class App {
  #persistenceRequest: Promise<boolean> | undefined;
  saveData() {
    this.#persistenceRequest ??= navigator.storage.persist();
  }
}
```

```ts
class App {
  #persistenceRequest: Promise<boolean> | undefined;
  saveData() {
    this.#persistenceRequest = this.#persistenceRequest  ?? navigator.storage.persist();
  }
}
```

as Valid (which differs from eslint themselves and useReadonlyClassProperties)

**Comment by Conaclos** [comment]
@lgarron I really don't get your point. Is your point about understanding the rule logic?

To me `a ??= b` can be considered different from `a = a ?? b`. The former can be seen as atomic.
Even if `a` is read, it looks fine to me to report this case as unused if `a` is not read anywhere else. Why do you want to keep a property that only read to update itself?

**Comment by vladimir-ivanov** [comment]
it is what useReadonlyClassProperties does @Conaclos , hence my other proposal to align the two rules.  https://github.com/biomejs/biome/issues/7499
My opinion as I stated above is a ??= b and a = a ?? b should be treated as equal though, whether both valid or both invalid.

Also useReadonlyClassProperties separates the logic depending whether the update  happens inside a constructor or elsewhere.

**Comment by vladimir-ivanov** [comment]
@Conaclos ok I would agree with you that AST nodes are slightly different 
prop ??= 3 is a compound assignment operator.
	•	It’s atomic: the engine checks whether prop is null/undefined, and if so, assigns 3.
	•	The read + write happen as a single low-level operation — you can’t observe the intermediate prop value separately.
	•	prop = prop ?? 3 is a desugared expression.
	•	It explicitly reads prop (to evaluate the ??).
	•	Then it writes the result back into prop.
	•	Conceptually equivalent in result, but from an AST/analysis perspective, it’s a real read followed by a write.

It is behaviour across other linters (not just eslint, but other languages too it seems) - C, Java, C#, Swift, Kotlin, Rust, etc.

@lgarron does it make sense to you please? if so we can close the ticket?

**Comment by vladimir-ivanov** [comment]
adding more details regarding the rules to #7499

**Comment by lgarron** [comment]
> [@lgarron](https://github.com/lgarron) does it make sense to you please? if so we can close the ticket?

I've gotta be honest, I've disengaged a bit from this issue because I don't see how this can be considered anything other than a false positive when it comes to "using" the value. I would be disappointed if the linter did not accept `??=` as a read (and therefore usage) of the left-hand side. That is literally how I use it in my code, and so far it has always been a false positive to consider it otherwise.

A few unfiltered thoughts:

> • It’s atomic: the engine checks whether prop is null/undefined, and if so, assigns 3. 

"Atomic" doesn't make any sense here.

- The semantics here concern only a single thread.
  - JS has [atomics](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Atomics), but that concerns something very different.
- It's possible to perform arbitrary work (even blocking on a network request) between the read and write, so I don't think "atomic" comes even close to describing it.

> The read + write happen as a single low-level operation — you can’t observe the intermediate prop value separately.

I mean, they don't happen as a single low-level operation. I also don't understand why a "separate observation" should affect anything.

> • prop = prop ?? 3 is a desugared expression.
> • It explicitly reads prop (to evaluate the ??).
> • Then it writes the result back into prop.

Sure.

> • Conceptually equivalent in result, but from an AST/analysis perspective, it’s a real read followed by a write.

I don't follow this at all.

**Comment by vladimir-ivanov** [comment]
cool, sounds like your concern is about side effect fn rather than the actual rule(s). 
My stake was to make things consistent In one centralised place first and then deal with bugs in one place.
navigator.storage.persist(); this can not be easily determined as a side/ non side effect fn in a listing tool. 

Will unassigned myself from it, and leave to someone with more powers than me to make a call. Thank you
